Multi-hop `BlindedPath` creation interface

jkczyz opened the pull request:
Expands `Router` and `MessageRouter` traits with methods for constructing `BlindedPath`s for payment and onion messaging, respectively. Also makes `MessageRouter` a super trait of `Router` so that `ChannelManager` doesn't need another parameter.

`ChannelManager` uses the new interface in its Offers flow to construct:
- paths for `Offers` and `Refunds`,
- reply paths for `InvoiceRequest` and `Bolt12Invoice`, and
- payment paths for `Bolt12Invoice`.

Each trait provides a default implementation of the new methods, which simply construct two-hop blinded paths (i.e., where the peer is the introduction node).

Resolves #2690

jkczyz commented:
@TheBlueMatt @valentinewallace Looking for high-level feedback on the new interface and how it is used by `ChannelManager`. I didn't implement the new methods for `DefaultRouter` or `DefaultMessageRouter` just yet. Instead, I simply provided trait-level defaults returning the naive two-hop blinded paths.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -270,6 +270,24 @@ pub trait MessageRouter {
+	fn create_blinded_paths<
+		ES: EntropySource + ?Sized, T: secp256k1::Signing + secp256k1::Verification
+	>(
+		&self, recipient: PublicKey, peers: Vec<PublicKey>, count: usize, entropy_source: &ES,
```
`Router:: create_blinded_payment_paths` is used for `Invoice` paths, so presumably the implementation could decide, yeah.

`Message::create_blinded_paths` as of now just needs one as mentioned above. If we allow implementations to decide, are you saying if an implementation returns more than one then we should set all of them on an `Offer` or `Refund` (where QR code size may be a factor)? For reply paths, would we simply use an arbitrary (first?) one? Or even cycle through them in case of sending multiple `InvoiceRequest`s?

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -105,6 +130,48 @@ pub trait Router {
+		&self, recipient: PublicKey, first_hops: Vec<ChannelDetails>, tlvs: ReceiveTlvs,
+		amount_msats: u64, count: usize, entropy_source: &ES, secp_ctx: &Secp256k1<T>
+	) -> Result<Vec<(BlindedPayInfo, BlindedPath)>, ()> {
+		first_hops.into_iter()
```
Ok, just noting others will need to implement this if they use their own `Router` implementation, possibly just wrapping `DefaultRouter` assuming they have a `NetworkGraph` on hand.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,80 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+					&[forward_node], recipient, tlvs.clone(), u64::MAX, entropy_source, secp_ctx
+				)
+			})
+			.take(MAX_PAYMENT_PATHS)
```
Chose to limit to three paths in the implementation for `DefaultRouter`. I think we want _some_ limit to avoid computing too many `BlindedPath`s.

jkczyz replied:
Wondering if should increase this substantially to avoid failed payments. Though, maybe the best strategy is to sort ascending all possible paths by scorer penalty and take some N.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -321,6 +335,20 @@ where
+		peers.into_iter()
+			.filter(|pubkey| network_graph.node(&NodeId::from_pubkey(pubkey)).is_some())
+			.map(|hop| vec![hop, recipient])
+			.map(|node_pks| BlindedPath::new_for_message(&node_pks, entropy_source, secp_ctx))
```
I guess we should limit this as well?

jkczyz commented on lightning/src/ln/features.rs:
```diff
@@ -1068,7 +1068,7 @@ mod tests {
 			// - option_data_loss_protect (req)
 			// - var_onion_optin (req) | static_remote_key (req) | payment_secret(req)
-			// - basic_mpp | wumbo | anchors_zero_fee_htlc_tx
+			// - basic_mpp | wumbo | option_anchors_zero_fee_htlc_tx
```
Mostly was aiming for internal consistency here and with BOLT 9, though I get why we don't use `option` in the identifier names.

jkczyz reviewed (commented):
PTAL. I think expanding the number of hops by one shouldn't be too difficult.

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -129,6 +130,48 @@ pub trait Router: MessageRouter {
+			.filter(|details| amount_msats >= details.inbound_htlc_minimum_msat.unwrap_or(0))
+			.filter(|details| amount_msats <= details.inbound_htlc_maximum_msat.unwrap_or(0))
+			.map(|details| {
+				let short_channel_id = details.get_inbound_payment_scid().unwrap();
```
I don't think we should unwrap here and below.

jkczyz replied:
IIUC, a usable channel will always have these set. I could add debug assertions or additional filters to be safe, if you prefer.

valentinewallace replied:
Yeah, while unlikely this panic is reachable from the public interface so I think we should filter.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7800,8 +7810,39 @@ where
 
-	/// Creates a one-hop blinded path with [`ChannelManager::get_our_node_id`] as the introduction
-	/// node.
+	/// Creates `count` blinded paths by delegating to [`MessageRouter::create_blinded_paths`].
+	///
+	/// May return fewer paths if there are fewer than `count` peers that [support onion messages].
```
Outdated docs.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7823,6 +7864,34 @@ where
+	/// Creates multi-hop blinded payment paths for the given `amount_msats` by delegating to
+	/// [`Router::create_blinded_payment_paths`].
+	///
+	/// May return no paths if no peers [`support route blinding`] or whose channels don't have
```
nit: remove ticks

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,80 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
 		)
 	}
+
+	fn create_blinded_payment_paths<
```
Given we have the amount, we probably want to use something more like `lightning-invoice`'s `sort_and_filter_channels` for this method, at least eventually. Currently the sum of inbound capacity of the paths we're returning may not be sufficient, IIUC.

jkczyz replied:
> Given we have the amount, we probably want to use something more like `lightning-invoice`'s `sort_and_filter_channels` for this method, at least eventually.

Hmm... perhaps eventually. Unlike in `lightning-invoice`, these are meant to be used immediately, so the channels should be live at very least when going to pay. We may want to prefer channels based on their available liquidity, which would require a `collect` and `sort`. I was trying to avoid this given this action can be initiated by anyone sending an onion message.

> Currently the sum of inbound capacity of the paths we're returning may not be sufficient, IIUC.

Added the check against inbound capacity for now as that was indeed missing. Currently, requiring that every channel has enough. I suppose if this constraint cannot be met we could relax it to a fraction of the amount based on `MAX_PAYMENT_PATHS`?

valentinewallace replied:
> I was trying to avoid this given this action can be initiated by anyone sending an onion message.

The action can be initiated by anyone only if this node has a public offer, is that right? 

Another thing to note is that if we include one of our channels that is public but doesn't have enough confirmations to be in the public network graph, we'll be revealing to our counterparty that we are the likely destination for the payment. This is something that `sort_and_filter_channels` takes care of, and based on the docs for the method it also takes other considerations into account that seem relevant here. This may not be a big concern for v1, though.

jkczyz replied:
> The action can be initiated by anyone only if this node has a public offer, is that right?

Yeah, true.

> 
> Another thing to note is that if we include one of our channels that is public but doesn't have enough confirmations to be in the public network graph, we'll be revealing to our counterparty that we are the likely destination for the payment.

Does `list_usable_channels` mostly avoid this? I believe `sort_and_filter_channels` doesn't assume all channels passed are usable. Wouldn't our counterparty know the payment is likely for us if the channel is unannounced as well?

> This is something that `sort_and_filter_channels` takes care of, and based on the docs for the method it also takes other considerations into account that seem relevant here. This may not be a big concern for v1, though.

Hmm... yeah there is some overlap there, but I think the scenarios are also different since the introduction nodes aren't our counterparties here whereas for `sort_and_filter_channels` the hints are. So the filtering and selection criteria aren't necessarily going to be the same as there.

Also, as mentioned elsewhere, we probably need to do some sorting after all so the payer knows which paths to prefer (e.g., ones where we have higher confidence that the hop from the introduction node has enough liquidity).

valentinewallace replied:
> Does list_usable_channels mostly avoid this? I believe sort_and_filter_channels doesn't assume all channels passed are usable. 

I believe `list_usable_channels` just checks that the channel is funded, not how many confirmations it has. 

> Wouldn't our counterparty know the payment is likely for us if the channel is unannounced as well?

Yeah, so if we have public and private channels we should prefer the public ones in blinded payment paths we return, IIUC. 

> So the filtering and selection criteria aren't necessarily going to be the same as there.

With you there, just saying we'll probably want to see which of its checks are relevant for us when we add sorting.

jkczyz replied:
> I believe `list_usable_channels` just checks that the channel is funded, not how many confirmations it has.

AFAICT, it goes it is consider usable when both parties sent `channel_ready`, which happens once the funding transaction has enough confirmations.

> Yeah, so if we have public and private channels we should prefer the public ones in blinded payment paths we return, IIUC.

I'm not sure I understand why. They are blinded from the payer, so the concern we have with invoice hints doesn't apply. Maybe our counterparty learns something if we have both public and private channels with them and the payer uses MPP over both? Let me know if I'm missing something.

valentinewallace replied:
> Maybe our counterparty learns something if we have both public and private channels with them and the payer uses MPP over both? Let me know if I'm missing something.

I was thinking that the counterparty can be fairly confident that we're the recipient if a private channel is used, but if public channels are used then we have more deniability. Privacy concerns aren't too high priority rn though, as discussed offline.

jkczyz reviewed (commented):
Added another hop to the blinded payment paths.

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,134 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+					}
+				};
+				BlindedPath::new_for_payment(
+					&[introduction_forward_node, counterparty_forward_node], recipient,
```
It seems potentially likely for users with some `LDKServer<>LDKClient` architecture that the counterparty will support route blinding, but none of their public counterparties do. I don't think we should fail to construct a path in that situation?

valentinewallace replied:
> How does the architecture play a role here?

The architecture/situation outlined seems like a somewhat common LDK use case where 3-hop wouldn't work, and 1-hop also wouldn't work (due to the client likely being private). 

To me it would make more sense if we always tried for 2-hop in this method, and fall back to 1-hop if the recipient is a public node, otherwise error. Then we wouldn't need to have 1-hop fallback logic in channelmanager as well. I may be missing something here, though.

jkczyz replied:
> > How does the architecture play a role here?
> 
> The architecture/situation outlined seems like a somewhat common LDK use case where 3-hop wouldn't work, and 1-hop also wouldn't work (due to the client likely being private).

Still a little confused about this. I thought you were talking about splitting an LDK node. But are you saying `LDKServer` is an LSP node and `LSPClient` is a separate node with a private channel with the LSP?

> To me it would make more sense if we always tried for 2-hop in this method, and fall back to 1-hop if the recipient is a public node, otherwise error. Then we wouldn't need to have 1-hop fallback logic in channelmanager as well. I may be missing something here, though.

When would we try 3-hops? Are you thinking something like this?

```rust
let paths = if has_public_channels {
    create_three_hop_paths()
} else {
    create_two_hop_paths()
};

if !paths.is_empty() {
    paths
} else if has_public_channels {
    create_one_hop_path()
} else {
    Err(())
}
```

jkczyz replied:
See https://github.com/lightningdevkit/rust-lightning/pull/2781#issuecomment-1855086265

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,134 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+	>(
+		&self, recipient: PublicKey, first_hops: Vec<ChannelDetails>, tlvs: ReceiveTlvs,
+		amount_msats: u64, entropy_source: &ES, secp_ctx: &Secp256k1<T>
+	) -> Result<Vec<(BlindedPayInfo, BlindedPath)>, ()> {
```
e2e would be possible with 2-hop blinded payment paths but not 3 :/

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,134 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+				)
+			})
+			.take(MAX_PAYMENT_PATHS)
+			.collect()
```
Seems like this should error if no paths are found?

jkczyz replied:
We treat these the same in `ChannelManager` by falling back to 1-hop paths. Not sure if we should move that code here, but failing when we don't have public channels means we can't send an invoice, FWIW.

jkczyz replied:
Moved fallback code to the respective default implementations, FYI.

jkczyz commented:
> i kinda wonder if we shouldn't max out at 2 hops for now. Its not great for privacy, but until we use scoring data in our blinding selection logic i fear we're just gonna always fail to receive, which is just gonna kill B12 adoption until we fix it.

Discussed offline. We'll use 2-hop paths for now and do 3-hop paths in another PR. Moved relevant commits to #2793 and squashed.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7800,8 +7810,39 @@ where
+
+	/// Creates a one-hop blinded payment path with [`ChannelManager::get_our_node_id`] as the
+	/// introduction node.
 	fn create_one_hop_blinded_payment_path(
```
So now that this is moved to `Router` (and `MessageRouter`, respectively), we can check there but only in `Router`. In `MessageRouter`, we don't have channel information.

jkczyz replied:
Oh, actually, we can just use the `NetworkGraph` to see if our node was announced.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,81 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+				let payment_relay: PaymentRelay = details.counterparty.forwarding_info.unwrap().into();
+				let payment_constraints = PaymentConstraints {
+					max_cltv_expiry: tlvs.payment_constraints.max_cltv_expiry
+						+ payment_relay.cltv_expiry_delta as u32,
```
Done. Let me know if that is what you were thinking.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -99,7 +99,6 @@ impl<G: Deref<Target = NetworkGraph<L>> + Clone, L: Deref, S: Deref, SP: Sized,
 
 		let network_graph = self.network_graph.deref().read_only();
 		let paths = first_hops.into_iter()
-			.filter(|details| details.is_public)
```
Needed to drop this since we want two-hop paths that contain an unannounced channel between the intro node and us.

jkczyz commented:
I wrote a test in #2697, but I'm getting the following failure when making the payment (debugging added):

```
[lightning/src/ln/onion_payment.rs:46] inbound_amt_msat = 10001000
[lightning/src/ln/onion_payment.rs:46] payment_constraints.htlc_minimum_msat = 1000
[lightning/src/ln/onion_payment.rs:46] dbg!(inbound_amt_msat) < dbg!(payment_constraints.htlc_minimum_msat) = false
[lightning/src/ln/onion_payment.rs:47] outgoing_cltv_value = 259
[lightning/src/ln/onion_payment.rs:47] payment_constraints.max_cltv_expiry = 93
[lightning/src/ln/onion_payment.rs:47] dbg!(outgoing_cltv_value) > dbg!(payment_constraints.max_cltv_expiry) = true
node 1 INFO [lightning::ln::onion_payment:393]         Failed to accept/forward incoming HTLC: Underflow calculating outbound amount or cltv value for blinded forward
```

So the CLTV check is causing the failure: https://github.com/jkczyz/rust-lightning/blob/181e621d26f9454c1e4726b7a923b9c8e26addf2/lightning/src/ln/onion_payment.rs#L47

Can be reproduced on that PR using:

```
cargo test -p lightning -- creates_and_pays_for_offer_using_two_hop_blinded_path
```

@valentinewallace Do you see anything wrong in the test setup or how I'm calculating the payment constraints? Note that I haven't rebased on your current work yet, so maybe that is it?

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7792,22 +7817,34 @@ where
 
+		let first_hops = self.list_usable_channels();
 		let payee_node_id = self.get_our_node_id();
 		let max_cltv_expiry = self.best_block.read().unwrap().height() + LATENCY_GRACE_PERIOD_BLOCKS;
```
Looks like this doesn't allow for the router adding the shadow offset to the CLTV during route construction. Will probably want to add `MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET` (from router.rs) to this value.

jkczyz replied:
Hmm... but the `BlindedPath` is created by the recipient, and the `Route` is created by the payer. Is that value common across implementations? Or just something we expect to be good enough?

valentinewallace replied:
True. I think it will be good enough, but this does mean that our router should be taking `max_cltv_expiry` into account and currently isn't :/

valentinewallace replied:
> Looks like this doesn't allow for the router adding the shadow offset to the CLTV during route construction. Will probably want to add `MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET` (from router.rs) to this value.

Thinking more, currently the way we set `max_cltv_expiry`here  doesn't allow for ~any CLTV delta along the route, not just the shadow offset. So I think the value we set here should increase even further, actually.

valentinewallace replied:
> True. I think it will be good enough, but this does mean that our router should be taking `max_cltv_expiry` into account and currently isn't :/

Ok took me a minute to reload all the context here. `max_cltv_expiry` isn't part of `BlindedPayInfo` so there's no way to take it into account when pathfinding. This confused me originally so I opened https://github.com/lightning/bolts/pull/1069/ to clarify it. But I don't think we should stop using the shadow offset in pathfinding just because invoice expiries tend to be short, so I'm not exactly sure what the resolution is here.

jkczyz replied:
Discussed offline to add `CLTV_FAR_FAR_AWAY`.

valentinewallace commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,109 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+		let paths = first_hops.into_iter()
+			.filter(|details| details.counterparty.features.supports_route_blinding())
+			.filter(|details| amount_msats <= details.inbound_capacity_msat)
+			.filter(|details| amount_msats >= details.inbound_htlc_minimum_msat.unwrap_or(0))
+			.filter(|details| amount_msats <= details.inbound_htlc_maximum_msat.unwrap_or(0))
```
IIUC, these should use `next_outbound_htlc_limit_msat`/`next_outbound_htlc_minimum_msat` because those are for routing purposes vs constant channel limits.

valentinewallace replied:
Ah, right. 🤦‍♀️

jkczyz replied:
But the blinded path is for inbound HTLCs. The recipient is constructing the blinded path for the payer to use to pay them. Maybe we need something similar for inbound, but it seems it would just an adjustment of `inbound_capacity_msat`, IIUC.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7792,22 +7817,34 @@ where
 
+		let first_hops = self.list_usable_channels();
 		let payee_node_id = self.get_our_node_id();
 		let max_cltv_expiry = self.best_block.read().unwrap().height() + LATENCY_GRACE_PERIOD_BLOCKS;
```
This should also account for the invoice expiry, IIUC. We default to 60 minutes right now, so I think we should add 6 blocks on for that (assuming 10 minute blocks), per https://github.com/lightning/bolts/pull/1069.

jkczyz replied:
Discussed offline to add `CLTV_FAR_FAR_AWAY`.

jkczyz commented:
Rebased and squashed.

jkczyz commented on lightning/src/routing/router.rs:
```diff
@@ -78,10 +83,110 @@ impl< G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref, SP: Sized, Sc: Sco
+					None => return None,
+				};
+				let payment_relay: PaymentRelay = match details.counterparty.forwarding_info {
+					Some(forwarding_info) => forwarding_info.into(),
```
Will leave to a follow-up.
